File: intersection_of_2LL.java
// Time Complexity :O(m+n) m-number of nodes in headA n-number of nodes in headB
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :-
public class Solution {

    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if(headA==null || headB==null) return null;
        ListNode temp=headA;
        ListNode temp2=headB;
        int length=0,length2=0;
      while(temp!=null)
      {
        temp=temp.next;
        length++;
    
      }
      while(temp2!=null)
      {
        temp2=temp2.next;
        length2++;
    
      }
      while(length<length2)
      {
          headB=headB.next;
          length2--;
      }
     while(length2<length)
      {
          headA=headA.next;
          length--;
      }
     while(headA!=headB)
     {
      headA=headA.next;   
      headB=headB.next;
        
     }
     return headA;

    }
}
//BruteForce
/*
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
       HashSet<ListNode> map=new  HashSet<ListNode>();
        while(headA!=null)
        {
            map.add(headA);
            headA=headA.next;
            
        }
        while(headB!=null)
        {
            if(map.contains(headB))
                return headB;
            headB=headB.next;
        }
        
       
        return null;
    }
}
*/

File: IntersectionOf2LL.cpp
// Brute-force solution using a set.
// Time Complexity - O(m+n) - Since we are iterating over 'm' nodes in list 'A' and 'n' nodes in List 'B'.
// Space Complexity - O(m) - Since we are using a set to store all the nodes in List 'A'.
// Problems Faced - No!
// It runs on leetcode.
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        set<ListNode*> s;
        ListNode* temp = headA;
        while(temp){
            s.insert(temp);
            temp = temp->next;
        }
        
        ListNode* tempB = headB;
        while(tempB){
            if(!s.count(tempB))
                tempB = tempB->next;
            else break;
        }
        return tempB;
    }
};

// Brute Force Solution without using a set.
// Time Complexity - O(m+n) - Since we are iterating over 'm' nodes in list 'A' and 'n' nodes in List 'B'.
// Space Complexity - O(1) - Since we are just using a pointer and variables to maintain count for the soultion.
// Problems Faced - No!
// It runs on leetcode.
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        int countA = 0;
        int countB = 0;
        
        ListNode* temp = headA;
        while(temp){
            countA++;
            temp = temp->next;
        }
        
        temp = headB;
        while(temp){
            countB++;
            temp = temp->next;
        }
        
        while(countA > countB){
            headA = headA->next;
            countA--;
        }
        
        while(countA < countB){
            headB = headB->next;
            countB--;
        }
        
        while(headA != headB){
            headA = headA->next;
            headB = headB->next;
        }
       return headA;
    }
};


// Approach 3
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        ListNode* currA = headA;
        ListNode* currB = headB;
        
        while(currA != currB){
            if(currA)
                currA = currA->next;
            else
                currA = headB;
            
            if(currB)
                currB = currB->next;
            else
                currB = headA;
        }
        return currA;
    }
};